src/__tests__/core/grammar.spec.ts

```diff
@@ -0,0 +1,24 @@
+import { grammar, run } from '@core'
+import { Parser, Success, choice, map, optional, sequence, string } from '@lib'
+import { describe, should, it } from '@testing'
+
+describe(`grammar`, () => {
+  it('should create self-contained grammar', () => {
+    const ab = grammar({
+      a(): Parser<string> {
+        return map(sequence(string('a'), optional(this.b)), ([a, b]) => a + (b || ''))
+      },
+      b(): Parser<string> {
+        return map(sequence(string('b'), choice(this.a, this.b)), ([b, a]) => b + a)
+      }
+    })
+
+    const value = 'abba'
+
+    const actual = run(ab.a).with(value)
+
+    const expected: Success<string> = { isOk: true, value, span: [0, 4], pos: 4 }
+
+    should.beStrictEqual(actual, expected)
+  })
+})
```

Comment:
The test can be simplified a bit:

```suggestion
import { grammar } from '@core'
import { choice, map, optional, sequence, string } from '@lib'
import { describe, should, it, run, result } from '@testing'

describe('grammar', () => {
  it('should create self-contained grammar', () => {
    const ab = grammar({
      a() {
        return map(sequence(string('a'), optional(this.b)), ([a, b]) => a + (b ?? ''))
      },
      b() {
        return map(sequence(string('b'), choice(this.a, this.b)), ([b, a]) => b + a)
      }
    })

    const value = 'abba'

    const actual = run(ab.a, value)
    const expected = result(true, value)

    should.matchState(actual, expected)
  })
})
```